src/freenet/support/io/InetAddressIpv6FirstComparator.java

```diff
@@ -0,0 +1,92 @@
+package freenet.support.io;
+
+import static freenet.node.NodeStats.DEFAULT_MAX_PING_TIME;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.util.Comparator;
+
+import freenet.support.Fields;
+import freenet.support.LRUCache;
+
+/** Comparator for IP addresses that sorts IPv6 before IPv4 to enable
+ * selecting the first.
+ * @author toad */
+public class InetAddressIpv6FirstComparator implements Comparator<InetAddress> {
+
+	// need a cache for reachability to avoid doing NlogN issReachable checks in worst case.
+	public LRUCache<Integer, Boolean> reachabilityCache = new LRUCache<>(1000, 300000);
+
+	public static final InetAddressIpv6FirstComparator COMPARATOR =
+		new InetAddressIpv6FirstComparator();
+
+	@Override
+	public int compare(InetAddress arg0, InetAddress arg1) {
+		if(arg0 == arg1) return 0;
```

Comment:
Is this a reference comparison? Or it is implicitly calling Equals()?

Reply:
it looks like a reference comparison, yes … damn … I’ll fix that during merging. Thank you!